# Association Rule Learning Recommendation System

Birliktelik Kuralına ilişkin olarak bir kullanıcının okuduğu kitaplara ilişkin en yakın hangi kitapları tavsiye edebiliriz.

### Importing Necessary Libraries

In [1]:
#pip install mlxtend

In [2]:
import numpy as np
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

pd.set_option('display.max_columns', 50)
pd.set_option('display.width', 50)
import warnings
warnings.filterwarnings('ignore')

### Reading Datasets

In [3]:
df= pd.read_csv('df_final_21_09___9794.csv',index_col=0)
df

book_id  \
0           1   
1           2   
2           3   
3           4   
4           5   
...       ...   
9789     9996   
9790     9997   
9791     9998   
9792     9999   
9793    10000   

                                             book_title  \
0                                      The Hunger Games   
1              Harry Potter and the Philosopher's Stone   
2                                              Twilight   
3                                 To Kill a Mockingbird   
4                                      The Great Gatsby   
...                                                 ...   
9789                                         Bayou Moon   
9790                                   Means of Ascent    
9791                              The Mauritius Command   
9792  Cinderella Ate My Daughter: Dispatches from th...   
9793                                The First World War   

                          book_series  \
0                The Hunger Games, #1   
1                    Harry Potter, #1   
2                        Twilight, #1   
3                                   -   
4                                   -   
...                               ...   
9789                     The Edge, #2   
9790  The Years of Lyndon Johnson, #2   
9791                                -   
9792                                -   
9793                                -   

                                                  title  \
0               The Hunger Games (The Hunger Games, #1)   
1     Harry Potter and the Sorcerer's Stone (Harry P...   
2                               Twilight (Twilight, #1)   
3                                 To Kill a Mockingbird   
4                                      The Great Gatsby   
...                                                 ...   
9789                          Bayou Moon (The Edge, #2)   
9790  Means of Ascent (The Years of Lyndon Johnson, #2)   
9791                              The Mauritius Command   
9792  Cinderella Ate My Daughter: Dispatches from th...   
9793                                The First World War   

                     book_authors  \
0                 Suzanne Collins   
1     J.K. Rowling, Mary GrandPré   
2                 Stephenie Meyer   
3                      Harper Lee   
4             F. Scott Fitzgerald   
...                           ...   
9789                Ilona Andrews   
9790               Robert A. Caro   
9791              Patrick O'Brian   
9792              Peggy Orenstein   
9793                  John Keegan   

                                                 genres  \
0     Young Adult|Fiction|Science Fiction|Dystopia|F...   
1                                                   NaN   
2     Young Adult|Fantasy|Romance|Paranormal|Vampire...   
3     Classics|Fiction|Historical|Historical Fiction...   
4     Classics|Fiction|Academic|School|Literature|Hi...   
...                                                 ...   
9789  Fantasy|Urban Fantasy|Fantasy|Romance|Fantasy|...   
9790                                                NaN   
9791  Historical|Historical Fiction|Fiction|Historic...   
9792  Nonfiction|Parenting|Feminism|Psychology|Socio...   
9793                                                NaN   

        book_score  book_rating  \
0     1.000000e+00         4.33   
1     8.245693e-01         4.44   
2     2.902011e-02         3.58   
3     2.310594e-01         4.27   
4     4.942671e-02         3.90   
...            ...          ...   
9789  5.206229e-06         4.10   
9790  1.744859e-06         4.25   
9791  8.571567e-07         4.35   
9792  1.661903e-06         3.66   
9793  6.551884e-07         4.00   

      book_rating_obj  book_rating_count  \
0                 4.0            5519135   
1                 4.0            6682936   
2                 3.5            4281268   
3                 4.0            3745197   
4                 3.5            3141842   
...               ...                ...   
9789   

In [4]:
df_read = pd.read_csv('to_read.csv',index_col=0)
df_read=df_read.reset_index()
df_read.head(5)

user_id  book_id
0        9        8
1       15      398
2       15      275
3       37     7173
4       34      380

In [5]:
df_ratings = pd.read_csv('ratings.csv',index_col=0)
df_ratings=df_ratings.reset_index()
df_ratings

user_id  book_id  rating
0              1      258       5
1              2     4081       4
2              2      260       5
3              2     9296       5
4              2     2318       3
...          ...      ...     ...
5976474    49925      510       5
5976475    49925      528       4
5976476    49925      722       4
5976477    49925      949       5
5976478    49925     1023       4

[5976479 rows x 3 columns]

### Data Preprocessing

In [6]:
df_ratings["user_id"].value_counts()

12874    200
30944    200
52036    199
12381    199
28158    199
        ... 
32128     21
40753     21
51725     21
43675     20
34590     19
Name: user_id, Length: 53424, dtype: int64

Veri setinde 53424 farklı tipte kullanıcı mevcut. Bu kullanıcılar arasında belli sayının üzerindeki kullanıcıları seçerek veri setini indirgeriz ve daha efektif sonuçlar elde ederiz.

> **Burada tüm kullanıcıları ve kitapları seçmemiz gerekiyor. Ancak bunu yapmamız durumunda yaklaşık 800 milyar satırlık-100 GB üstü veri elde ediliyor. Bu veriyi bu ortamlarda elde etmemiz çok uzun sürüyor vede bu kadar alanıımız yok. bu nedenle bu kullanıcı ve kitap sayısını optimum sayıya azaltarak sonuçlarımızı göstermek istedik.Bu işlemler sonucunda 20 üzeri kitap okuyan tüm kullanıcılar için en beğenilen 500 kitap arasından önerilerde bulunmuş olduk.**

>20 üzeri kitap okuyan kullanıcıları seçtik

In [7]:
df_bookish_user=df_ratings["user_id"].value_counts()
df_bookish_user=df_bookish_user[df_bookish_user>20].index

>score değeri ortalama üzeri olan, 4.31 değer üzeri rating puanı olan 500 kitabı seçtik

In [8]:
df6=df[df["book_rating"]>4.31][df["book_score"]>df["book_score"].median()]
df6=df6.iloc[:500]
df_book_ids=df6["book_id"].values

Yukarıda oluşturduğumuz kullanıcıları ve kitapları daha önce tanımladığımız "df_read" veri setinde okutup yeni bir veri seti oluşturuyoruz.

In [9]:
df_read3=df_read[df_read["user_id"].isin(df_bookish_user)][df_read["book_id"].isin(df_book_ids)]

Yeni oluşturduğumuz "df_read3" veri setine kitap başlıklarımızı da ekleyip veri setini güncelliyoruz.

In [10]:
df_read3=df_read3.merge(df[["book_id","title"]],on="book_id") 
df_read3["Quantity"]=1
df_read3

user_id  book_id  \
0          207       25   
1          590       25   
2          656       25   
3          779       25   
4         1147       25   
...        ...      ...   
70526    19344     4641   
70527    31724     4641   
70528    29044     4641   
70529    27692     4641   
70530    47146     4641   

                                                   title  \
0      Harry Potter and the Deathly Hallows (Harry Po...   
1      Harry Potter and the Deathly Hallows (Harry Po...   
2      Harry Potter and the Deathly Hallows (Harry Po...   
3      Harry Potter and the Deathly Hallows (Harry Po...   
4      Harry Potter and the Deathly Hallows (Harry Po...   
...                                                  ...   
70526  Fantastic Beasts and Where to Find Them: The O...   
70527  Fantastic Beasts and Where to Find Them: The O...   
70528  Fantastic Beasts and Where to Find Them: The O...   
70529  Fantastic Beasts and Where to Find Them: The O...   
70530  Fantastic Beasts and Where to Find Them: The O...   

       Quantity  
0             1  
1             1  
2             1  
3             1  
4             1  
...         ...  
70526         1  
70527         1  
70528         1  
70529         1  
70530         1  

[70531 rows x 4 columns]

In [11]:
df_read3["user_id"].nunique()

27772

In [12]:
df_read3["book_id"].nunique()

500

> Yapılan işlem sonrasında **500 kitap ve 27772 kullanıcı**mız mevcuttur.

# Association Rule Learning(Birliktelik Kuralı Öğrenimi)

> Veritabanları içerisinde ürünler arasındaki ilişkiyi ortaya çıkarmak için kullanılan temel tekniklerden biridir. Başka bir deyişle, birlikte daha sık satın alınan ürünler arasındaki ilişkiyi belirlemeyi sağlar.

 >Bu kural şu adlandırmalar ile de adlandırılmıştır: Market Sepet Analizi (Market Basket Analysis), Birliktelik Kuralları (Association Rules), Tavsiye Sistemleri (Recommended Systems)

**Define  Matrix Function**

In [13]:
def create_matrix_title_df(dataframe, id=False):
    if id:
        return dataframe.groupby(['user_id', "book_id"])['Quantity'].sum().unstack().fillna(0). \
            applymap(lambda x: 1 if x > 0 else 0)
    else:
        return dataframe.groupby(['user_id', 'title'])['Quantity'].sum().unstack().fillna(0). \
            applymap(lambda x: 1 if x > 0 else 0)
# burada seçimimiz sonrası "book_title" veya 'book_id'a göre bize matriks oluşturacak

In [14]:
df_ar_title = create_matrix_title_df(df_read3)
df_ar_title

title    A Breath of Snow and Ashes (Outlander, #6)  \
user_id                                               
2                                                 0   
5                                                 0   
6                                                 0   
7                                                 0   
8                                                 0   
...                                             ...   
53417                                             0   
53421                                             0   
53422                                             0   
53423                                             0   
53424                                             0   

title    A Clash of Kings  (A Song of Ice and Fire, #2)  \
user_id                                                   
2                                                     0   
5                                                     0   
6                                                     0   
7                                                     0   
8                                                     0   
...                                                 ...   
53417                                                 0   
53421                                                 0   
53422                                                 0   
53423                                                 0   
53424                                                 0   

title    A Court of Mist and Fury (A Court of Thorns and Roses, #2)  \
user_id                                                               
2                                                        0            
5                                                        0            
6                                                        0            
7                                                        0            
8                                                        0            
...                                                    ...            
53417                                                    0            
53421                                                    0            
53422                                                    0            
53423                                                    0            
53424                                                    0            

title    A Court of Wings and Ruin (A Court of Thorns and Roses, #3)  \
user_id                                                                
2                                                        0             
5                                                        0             
6                                                        0             
7                                                        0             
8                                                        0             
...                                                    ...             
53417                                                    0             
53421                                                    0             
53422                                                    0             
53423                                                    0             
53424                                                    0             

title    A Dance with Dragons (A Song of Ice and Fire, #5)  \
user_id                                                      
2                                                        0   
5                                                        0   
6                                                        0   
7                                                        0   
8                                                        0   
...                                                    ...   
53417                                                    0   
53421                                                    0   
53422                              

**Kullanıcı bazında kitap başlıklarına göre matrix veri seti tanımlıyoruz.**

> book_id sini bildiğimiz kitabın adını getiren fonksiyon tanımladık

In [15]:
def check_id(dataframe, book_id):
    product_name = dataframe[dataframe["book_id"] == book_id][["title"]].values[0].tolist()
    print(product_name)


check_id(df, 1)

['The Hunger Games (The Hunger Games, #1)']


## Birliktelik Kurallarının Çıkarılması

* 'antecedents',        ilk ürün
* 'consequents',        ikinci ürün
* 'antecedent support', ürünün tek başına gözlenme olasılığı
* 'consequent support', ürünün tek başına gözlenme olasılığı
* 'support',            ürünlerin birlikte gözlenme olasılığı
* 'confidence',         X alındığında Y'nin alınma olasılığı
* 'lift',               X alındığında Y'nin alınma olasılığı ... kat artar
* 'leverage',     kaldıraç etkisi demek. lift'e benzer bir değerdir Fakat sapportu yüksek olanlara öncelik verme eğilimindedir. Bu nedenle biraz yanlıdır. Lift ise daha az sıklıkta olmasına rağmen ilişkileri yakalayabilmektedir. 
* 'conviction'    Y/X ürünü olmadan X/Y'in beklenen frekansıdır

In [16]:
# min_support=0.002 değerine göre seçtiğimiz kitaplar arasında anlamlı ilişki olan 553 kitap buldu.
# daha büyük RAM ve database ortamında min_support değeri küçültülerek daha büyük ilişki ağı ortaya çıkarılabilir. 
# burada en önemli olan 553 ilişkiyi tespit etti.

# Apriori ilkesi, destek için minimum eşik koşulunu karşılamayan bir öğe kümesinin tüm üst kümelerini budamamıza izin verir. 
# Örneğin, {A, B}, minimum destek eşiğimizi karşılamıyorsa, 
# buna herhangi bir öğe eklenmiş bir öğe kümesi de eşiği asla geçmeyecektir.

In [17]:
frequent_itemsets = apriori(df_ar_title,
                            min_support=0.002,
                            use_colnames=True)

frequent_itemsets.sort_values("support", ascending=False)

support  \
166  0.099813   
20   0.070827   
7    0.058296   
199  0.051023   
17   0.048682   
..        ...   
295  0.002016   
235  0.002016   
413  0.002016   
523  0.002016   
298  0.002016   

                                              itemsets  
166                                   (The Book Thief)  
20                       (All the Light We Cannot See)  
7     (A Game of Thrones (A Song of Ice and Fire, #1))  
199                                         (The Help)  
17                          (A Thousand Splendid Suns)  
..                                                 ...  
295  (All the Light We Cannot See, Between Shades o...  
235  (The War that Saved My Life (The War That Save...  
413  (Dragonfly in Amber (Outlander, #2), Written i...  
523  (Drums of Autumn (Outlander, #4), An Echo in t...  
298  (All the Light We Cannot See, Six of Crows (Si...  

[553 rows x 2 columns]

In [18]:
# frequent_itemsets öğe kümelerinden tüm olası kuralları oluşturma

In [19]:
rules = association_rules(frequent_itemsets,
                          metric="support",
                          min_threshold=0.000001)
rules


antecedents  \
0                 (An Echo in the Bone (Outlander, #7))   
1          (A Breath of Snow and Ashes (Outlander, #6))   
2                  (Dragonfly in Amber (Outlander, #2))   
3          (A Breath of Snow and Ashes (Outlander, #6))   
4                     (Drums of Autumn (Outlander, #4))   
...                                                 ...   
2639  (Harry Potter and the Half-Blood Prince (Harry...   
2640  (Harry Potter and the Order of the Phoenix (Ha...   
2641  (Harry Potter and the Prisoner of Azkaban (Har...   
2642  (Harry Potter and the Chamber of Secrets (Harr...   
2643  (Harry Potter and the Goblet of Fire (Harry Po...   

                                            consequents  \
0          (A Breath of Snow and Ashes (Outlander, #6))   
1                 (An Echo in the Bone (Outlander, #7))   
2          (A Breath of Snow and Ashes (Outlander, #6))   
3                  (Dragonfly in Amber (Outlander, #2))   
4          (A Breath of Snow and Ashes (Outlander, #6))   
...                                                 ...   
2639  (Harry Potter and the Sorcerer's Stone (Harry ...   
2640  (Harry Potter and the Sorcerer's Stone (Harry ...   
2641  (Harry Potter and the Sorcerer's Stone (Harry ...   
2642  (Harry Potter and the Sorcerer's Stone (Harry ...   
2643  (Harry Potter and the Sorcerer's Stone (Harry ...   

      antecedent support  consequent support  \
0               0.010262            0.009866   
1               0.009866            0.010262   
2               0.011234            0.009866   
3               0.009866            0.011234   
4               0.009254            0.009866   
...                  ...                 ...   
2639            0.015771            0.002449   
2640            0.015375            0.002521   
2641            0.015411            0.002485   
2642            0.014691            0.002485   
2643            0.015519            0.002485   

       support  confidence       lift  leverage  \
0     0.006733    0.656140  66.504853  0.006632   
1     0.006733    0.682482  66.504853  0.006632   
2     0.003313    0.294872  29.887516  0.003202   
3     0.003313    0.335766  29.887516  0.003202   
4     0.005221    0.564202  57.186231  0.005130   
...        ...         ...        ...       ...   
2639  0.002376    0.150685  61.541499  0.002338   
2640  0.002376    0.154567  61.323252  0.002338   
2641  0.002376    0.154206  62.066640  0.002338   
2642  0.002376    0.161765  65.109122  0.002340   
2643  0.002376    0.153132  61.634621  0.002338   

      conviction  
0       2.879471  
1       3.117105  
2       1.404190  
3       1.488581  
4       2.272004  
...          ...  
2639    1.174536  
2640    1.179844  
2641    1.179383  
2642    1.190018  
2643    1.177888  

[2644 rows x 9 columns]

In [20]:
rules["antecedents"].nunique()

275

Burada birliktelik kuralına ilişkin olarak "Support", "Confidence" ve "Lift" değerlerini belli değerlerin üzerinde seçiyoruz ve "confidence" değerine göre büyükten küçüğe sıralıyoruz.

In [21]:
rules[(rules["support"]>0.002) & (rules["confidence"]>0.1) & (rules["lift"]>50)]. \
sort_values("confidence", ascending=False)

antecedents  \
833   (Dragonfly in Amber (Outlander, #2), Written i...   
1427  (Drums of Autumn (Outlander, #4), Written in M...   
1425  (Drums of Autumn (Outlander, #4), Written in M...   
804   (Drums of Autumn (Outlander, #4), An Echo in t...   
1365  (Drums of Autumn (Outlander, #4), A Breath of ...   
...                                                 ...   
2204  (Harry Potter and the Half-Blood Prince (Harry...   
2640  (Harry Potter and the Order of the Phoenix (Ha...   
2641  (Harry Potter and the Prisoner of Azkaban (Har...   
2643  (Harry Potter and the Goblet of Fire (Harry Po...   
2639  (Harry Potter and the Half-Blood Prince (Harry...   

                                            consequents  \
833               (An Echo in the Bone (Outlander, #7))   
1427              (An Echo in the Bone (Outlander, #7))   
1425                          (Voyager (Outlander, #3))   
804                           (Voyager (Outlander, #3))   
1365                          (Voyager (Outlander, #3))   
...                                                 ...   
2204  (Harry Potter and the Sorcerer's Stone (Harry ...   
2640  (Harry Potter and the Sorcerer's Stone (Harry ...   
2641  (Harry Potter and the Sorcerer's Stone (Harry ...   
2643  (Harry Potter and the Sorcerer's Stone (Harry ...   
2639  (Harry Potter and the Sorcerer's Stone (Harry ...   

      antecedent support  consequent support  \
833             0.002052            0.010262   
1427            0.002016            0.010262   
1425            0.002016            0.009542   
804             0.003061            0.009542   
1365            0.002809            0.009542   
...                  ...                 ...   
2204            0.015771            0.002557   
2640            0.015375            0.002521   
2641            0.015411            0.002485   
2643            0.015519            0.002485   
2639            0.015771            0.002449   

       support  confidence        lift  \
833   0.002052    1.000000   97.445614   
1427  0.002016    1.000000   97.445614   
1425  0.002016    1.000000  104.800000   
804   0.003025    0.988235  103.567059   
1365  0.002773    0.987179  103.456410   
...        ...         ...         ...   
2204  0.002485    0.157534   61.620297   
2640  0.002376    0.154567   61.323252   
2641  0.002376    0.154206   62.066640   
2643  0.002376    0.153132   61.634621   
2639  0.002376    0.150685   61.541499   

      leverage  conviction  
833   0.002031         inf  
1427  0.001996         inf  
1425  0.001997         inf  
804   0.002995   84.188931  
1365  0.002746   77.255725  
...        ...         ...  
2204  0.002444    1.183957  
2640  0.002338    1.179844  
2641  0.002338    1.179383  
2643  0.002338    1.177888  
2639  0.002338    1.174536  

[2064 rows x 9 columns]

In [22]:
sorted_rules = rules.sort_values("confidence", ascending=False)
# burada lift,confidence veya support' a göre sıralama yapılabilir. Hangisini önemli görüyor isek.
sorted_rules

antecedents  \
1425  (Drums of Autumn (Outlander, #4), Written in M...   
1427  (Drums of Autumn (Outlander, #4), Written in M...   
833   (Dragonfly in Amber (Outlander, #2), Written i...   
804   (Drums of Autumn (Outlander, #4), An Echo in t...   
1365  (Drums of Autumn (Outlander, #4), A Breath of ...   
...                                                 ...   
57                                     (The Book Thief)   
249                                    (The Book Thief)   
216                                    (The Book Thief)   
245                                    (The Book Thief)   
17                                     (The Book Thief)   

                                           consequents  \
1425                         (Voyager (Outlander, #3))   
1427             (An Echo in the Bone (Outlander, #7))   
833              (An Echo in the Bone (Outlander, #7))   
804                          (Voyager (Outlander, #3))   
1365                         (Voyager (Outlander, #3))   
...                                                ...   
57    (A Storm of Swords (A Song of Ice and Fire, #3))   
249                                           (Wonder)   
216                         (Man's Search for Meaning)   
245                                        (The Stand)   
17    (A Clash of Kings  (A Song of Ice and Fire, #2))   

      antecedent support  consequent support  \
1425            0.002016            0.009542   
1427            0.002016            0.010262   
833             0.002052            0.010262   
804             0.003061            0.009542   
1365            0.002809            0.009542   
...                  ...                 ...   
57              0.099813            0.020416   
249             0.099813            0.015771   
216             0.099813            0.017212   
245             0.099813            0.019732   
17              0.099813            0.021028   

       support  confidence        lift  \
1425  0.002016    1.000000  104.800000   
1427  0.002016    1.000000   97.445614   
833   0.002052    1.000000   97.445614   
804   0.003025    0.988235  103.567059   
1365  0.002773    0.987179  103.456410   
...        ...         ...         ...   
57    0.002232    0.022367    1.095526   
249   0.002160    0.021645    1.372433   
216   0.002088    0.020924    1.215665   
245   0.002052    0.020563    1.042097   
17    0.002052    0.020563    0.977858   

      leverage  conviction  
1425  0.001997         inf  
1427  0.001996         inf  
833   0.002031         inf  
804   0.002995   84.188931  
1365  0.002746   77.255725  
...        ...         ...  
57    0.000195    1.001995  
249   0.000586    1.006004  
216   0.000370    1.003791  
245   0.000083    1.000848  
17   -0.000046    0.999525  

[2644 rows x 9 columns]

In [23]:
# Kullanıcı A ve E kitaplarını okumuş olsun.

# Burada {A,B}-{C,D} kümeleri arasında ilişki var ise,
#    {A,C},
#    {A,D},
#    {B,C},
#    {B,D}  arasında ilişki vardır.

# ek olarak  {E,F}-{C,D} kümeleri arasında ilişki var ise,
#    {E,C},
#    {E,D},
#    {F,C},
#    {F,D}  arasında ilişki vardır.

# Bu ilişkileri ikinci sutun ile groupby yaptığımızda kullanıcının okuduğu A,E ile en ilişkili kitabın C olduğunu bulmayı 
# amaçlıyoruz.
# Bunu yapabilmek için kitaplar arasında tekli ilişkiyi gösteren aşağıdaki işlemi uyguluyoruz.

# İstenmesi durumunda bu çalışma ile standart yapılan çalışmalar gibi tek bir kitap için de tavsiyede bulunulabilir.

In [24]:
dfx=sorted_rules.copy()
cols = ['antecedents','consequents']
dfy=dfx[cols].applymap(lambda x: tuple(x))
dfy=pd.concat([dfy,dfx[dfx.columns[2:]]],axis=1)
dfy = (dfy.explode('antecedents')
         .reset_index(drop=True)
         .explode('consequents')
         .reset_index(drop=True))
dfy

antecedents  \
0                      Drums of Autumn (Outlander, #4)   
1      Written in My Own Heart's Blood (Outlander, #8)   
2                  An Echo in the Bone (Outlander, #7)   
3                   Dragonfly in Amber (Outlander, #2)   
4                      Drums of Autumn (Outlander, #4)   
...                                                ...   
11931                                   The Book Thief   
11932                                   The Book Thief   
11933                                   The Book Thief   
11934                                   The Book Thief   
11935                                   The Book Thief   

                                          consequents  \
0                             Voyager (Outlander, #3)   
1                             Voyager (Outlander, #3)   
2                             Voyager (Outlander, #3)   
3                             Voyager (Outlander, #3)   
4                 An Echo in the Bone (Outlander, #7)   
...                                               ...   
11931  A Storm of Swords (A Song of Ice and Fire, #3)   
11932                                          Wonder   
11933                        Man's Search for Meaning   
11934                                       The Stand   
11935  A Clash of Kings  (A Song of Ice and Fire, #2)   

       antecedent support  consequent support  \
0                0.002016            0.009542   
1                0.002016            0.009542   
2                0.002016            0.009542   
3                0.002016            0.009542   
4                0.002016            0.010262   
...                   ...                 ...   
11931            0.099813            0.020416   
11932            0.099813            0.015771   
11933            0.099813            0.017212   
11934            0.099813            0.019732   
11935            0.099813            0.021028   

        support  confidence        lift  \
0      0.002016    1.000000  104.800000   
1      0.002016    1.000000  104.800000   
2      0.002016    1.000000  104.800000   
3      0.002016    1.000000  104.800000   
4      0.002016    1.000000   97.445614   
...         ...         ...         ...   
11931  0.002232    0.022367    1.095526   
11932  0.002160    0.021645    1.372433   
11933  0.002088    0.020924    1.215665   
11934  0.002052    0.020563    1.042097   
11935  0.002052    0.020563    0.977858   

       leverage  conviction  
0      0.001997         inf  
1      0.001997         inf  
2      0.001997         inf  
3      0.001997         inf  
4      0.001996         inf  
...         ...         ...  
11931  0.000195    1.001995  
11932  0.000586    1.006004  
11933  0.000370    1.003791  
11934  0.000083    1.000848  
11935 -0.000046    0.999525  

[11936 rows x 9 columns]

In [25]:
# İlgili kitapların book_id değerlerini buluyoruz 

In [26]:
dfy=dfy.merge(df[["title","book_id"]], left_on="antecedents",right_on="title")
dfy = dfy.rename(columns={'book_id': 'book_id_antecedents'})

dfy=dfy.merge(df[["title","book_id"]], left_on="consequents",right_on="title")
dfy = dfy.rename(columns={'book_id': 'book_id_consequents'})

dfy=dfy[['book_id_antecedents', 'book_id_consequents','antecedents', 'consequents',
       'antecedent support','consequent support', 'support','confidence', 'lift', 'leverage','conviction' ]]
dfy

book_id_antecedents  book_id_consequents  \
0                      712                  554   
1                      712                  554   
2                      712                  554   
3                      712                  554   
4                      712                  554   
...                    ...                  ...   
11931                  143                  983   
11932                  143                 1192   
11933                   47                  250   
11934                   47                  472   
11935                   47                  168   

                           antecedents  \
0      Drums of Autumn (Outlander, #4)   
1      Drums of Autumn (Outlander, #4)   
2      Drums of Autumn (Outlander, #4)   
3      Drums of Autumn (Outlander, #4)   
4      Drums of Autumn (Outlander, #4)   
...                                ...   
11931      All the Light We Cannot See   
11932      All the Light We Cannot See   
11933                   The Book Thief   
11934                   The Book Thief   
11935                   The Book Thief   

                    consequents  \
0       Voyager (Outlander, #3)   
1       Voyager (Outlander, #3)   
2       Voyager (Outlander, #3)   
3       Voyager (Outlander, #3)   
4       Voyager (Outlander, #3)   
...                         ...   
11931  Between the World and Me   
11932    Between Shades of Gray   
11933                    Wonder   
11934  Man's Search for Meaning   
11935                 The Stand   

       antecedent support  consequent support  \
0                0.002016            0.009542   
1                0.003061            0.009542   
2                0.002809            0.009542   
3                0.002052            0.003817   
4                0.002052            0.009542   
...                   ...                 ...   
11931            0.070827            0.013791   
11932            0.070827            0.016023   
11933            0.099813            0.015771   
11934            0.099813            0.017212   
11935            0.099813            0.019732   

        support  confidence        lift  \
0      0.002016    1.000000  104.800000   
1      0.003025    0.988235  103.567059   
2      0.002773    0.987179  103.456410   
3      0.002016    0.982456  257.403509   
4      0.002016    0.982456  102.961404   
...         ...         ...         ...   
11931  0.002016    0.028470    2.064391   
11932  0.002016    0.028470    1.776768   
11933  0.002160    0.021645    1.372433   
11934  0.002088    0.020924    1.215665   
11935  0.002052    0.020563    1.042097   

       leverage  conviction  
0      0.001997         inf  
1      0.002995   84.188931  
2      0.002746   77.255725  
3      0.002009   56.782443  
4      0.001997   56.456107  
...         ...         ...  
11931  0.001040    1.015109  
11932  0.000882    1.012811  
11933  0.000586    1.006004  
11934  0.000370    1.003791  
11935  0.000083    1.000848  

[11936 rows x 11 columns]

In [27]:
# Burada birbiri arasında ilişki olan kitap grublarından birini görüyoruz

In [28]:
sorted_rules["antecedents"].iloc[0]

frozenset({'An Echo in the Bone (Outlander, #7)',
           'Dragonfly in Amber (Outlander, #2)',
           'Drums of Autumn (Outlander, #4)',
           "Written in My Own Heart's Blood (Outlander, #8)"})

In [29]:
sorted_rules["consequents"].iloc[0]

frozenset({'Voyager (Outlander, #3)'})

In [30]:
# bir kitabı okuyan kullanıcıları görüyoruz

In [31]:
df_read3[df_read3["title"]=="Clockwork Angel (The Infernal Devices, #1)"]

user_id  book_id  \
34642     7241      183   
34643     9430      183   
34644    10099      183   
34645     4011      183   
34646     8068      183   
...        ...      ...   
35494    50571      183   
35495    48721      183   
35496    50609      183   
35497    48384      183   
35498    24952      183   

                                            title  \
34642  Clockwork Angel (The Infernal Devices, #1)   
34643  Clockwork Angel (The Infernal Devices, #1)   
34644  Clockwork Angel (The Infernal Devices, #1)   
34645  Clockwork Angel (The Infernal Devices, #1)   
34646  Clockwork Angel (The Infernal Devices, #1)   
...                                           ...   
35494  Clockwork Angel (The Infernal Devices, #1)   
35495  Clockwork Angel (The Infernal Devices, #1)   
35496  Clockwork Angel (The Infernal Devices, #1)   
35497  Clockwork Angel (The Infernal Devices, #1)   
35498  Clockwork Angel (The Infernal Devices, #1)   

       Quantity  
34642         1  
34643         1  
34644         1  
34645         1  
34646         1  
...         ...  
35494         1  
35495         1  
35496         1  
35497         1  
35498         1  

[857 rows x 4 columns]

In [32]:
# Örnek olarak seçtiğimiz bir kitaba ait ilişki ağını görüyoruz

In [33]:
product_id = 25
check_id(df, product_id)

['Harry Potter and the Deathly Hallows (Harry Potter, #7)']


In [34]:
product_name=frozenset({'Harry Potter and the Deathly Hallows (Harry Potter, #7)'})

In [35]:
# sırasıyla tavsiye edilenler
recommendation_list=sorted_rules[sorted_rules["antecedents"]==product_name]#.iloc[:3,1]#.values
recommendation_list

antecedents  \
181   (Harry Potter and the Deathly Hallows (Harry P...   
183   (Harry Potter and the Deathly Hallows (Harry P...   
179   (Harry Potter and the Deathly Hallows (Harry P...   
582   (Harry Potter and the Deathly Hallows (Harry P...   
185   (Harry Potter and the Deathly Hallows (Harry P...   
...                                                 ...   
2513  (Harry Potter and the Deathly Hallows (Harry P...   
2389  (Harry Potter and the Deathly Hallows (Harry P...   
2203  (Harry Potter and the Deathly Hallows (Harry P...   
2327  (Harry Potter and the Deathly Hallows (Harry P...   
2638  (Harry Potter and the Deathly Hallows (Harry P...   

                                            consequents  \
181   (Harry Potter and the Half-Blood Prince (Harry...   
183   (Harry Potter and the Order of the Phoenix (Ha...   
179   (Harry Potter and the Goblet of Fire (Harry Po...   
582   (Harry Potter and the Half-Blood Prince (Harry...   
185   (Harry Potter and the Prisoner of Azkaban (Har...   
...                                                 ...   
2513  (Harry Potter and the Sorcerer's Stone (Harry ...   
2389  (Harry Potter and the Sorcerer's Stone (Harry ...   
2203  (Harry Potter and the Sorcerer's Stone (Harry ...   
2327  (Harry Potter and the Sorcerer's Stone (Harry ...   
2638  (Harry Potter and the Sorcerer's Stone (Harry ...   

      antecedent support  consequent support  \
181             0.017428            0.015771   
183             0.017428            0.015375   
179             0.017428            0.015519   
582             0.017428            0.010622   
185             0.017428            0.015411   
...                  ...                 ...   
2513            0.017428            0.002989   
2389            0.017428            0.002953   
2203            0.017428            0.002917   
2327            0.017428            0.002953   
2638            0.017428            0.002809   

       support  confidence       lift  leverage  \
181   0.011126    0.638430  40.480528  0.010851   
183   0.009686    0.555785  36.148160  0.009418   
179   0.009146    0.524793  33.815689  0.008875   
582   0.008822    0.506198  47.654714  0.008637   
185   0.008678    0.497934  32.309859  0.008409   
...        ...         ...        ...       ...   
2513  0.002485    0.142562  47.701583  0.002432   
2389  0.002485    0.142562  48.283310  0.002433   
2203  0.002485    0.142562  48.879400  0.002434   
2327  0.002449    0.140496  47.583552  0.002397   
2638  0.002376    0.136364  48.552448  0.002328   

      conviction  
181     2.722095  
183     2.216551  
179     2.071690  
582     2.003594  
185     1.961074  
...          ...  
2513    1.162780  
2389    1.162822  
2203    1.162864  
2327    1.160026  
2638    1.154643  

[63 rows x 9 columns]

# Kullanıcının okuduğu tüm kitapları dikkate alarak ARL ile kitap tavsiyesi

* ........................................şu kitaplara en yakın kitap bul (en iyi 500 kitap içerisinden)---denilerek tavsiye sistemi geliştirilebilir
* Buradan seçilen kitaplar  "okudugu_kitaplar" a atanır.Okudugu_kitaplar="Kullanıcının 'şu kitablara en yakın kitap bul' dan seçtiği kitapplar"


In [36]:
# Örnek olarak bir kullanıcı seçiyoruz

In [37]:
user=df_bookish_user[0]
user

12874

In [38]:
# Okuduğu kitapları buluyoruz

In [39]:
okudugu_kitaplar=df_ratings[df_ratings["user_id"]==user]["book_id"].values
okudugu_kitaplar

array([ 306,  118,   43,  943, 1186,  715,  138,  301,   93,  560,   79,
        341,   15,   66,   42,   75, 1457,   22,  142,   33,   57,    8,
         11,    4,    5,   24,   67,   21,  115,   26,   81,  407,    9,
        111,  331,  616,  216,  344,  383, 4793, 3406,  414,  465,  333,
       1685,  145, 1150,  955,  228, 1534,  150,   10,  265,   13,  131,
        501,   48,  233,  121,  916,   14,  230,  310,   63,  343,  178,
        650,  172,   55, 1714,  177,  171,  485,  396,  101,  714,  160,
        597,   83,   56,   85,   18,  133,  229,  117,  109,   29,  102,
         59,   58,  219,  125,   23,   76,  113,  179,  157,  498,   50,
        514,  130,  204,  321,  154,  116,   44,  296,  740, 1167,  213,
       1214,  444,  401,  427,  364,  264,  107,  959,  819,  358,  245,
        964,  984,  734,  877,  595,  278, 2142,  570,  621,  531, 2297,
        545,  471, 1099,  872, 1561, 1681,  479,  579,  402,  257, 1084,
       1015, 1280,  824, 2336,  497, 2236,    2,  2

In [40]:
# Okuduğu kitaplar ile ilişkisi olan ama daha önce hiç okumadığı kitaplar arasından, en büyük ilişki olan kitapları 
# önem sırasına göre sıralıyoruz. 

In [41]:
dfz=dfy[dfy["book_id_antecedents"].isin(okudugu_kitaplar)][~(dfy["book_id_consequents"].isin(okudugu_kitaplar))]
dfz=dfz.groupby("consequents").sum().sort_values("support",ascending=False)
dfz

book_id_antecedents  \
consequents                                                               
Harry Potter and the Half-Blood Prince (Harry P...                21384   
Harry Potter and the Deathly Hallows (Harry Pot...                21384   
The Fellowship of the Ring (The Lord of the Rin...                 1137   
The Book Thief                                                      256   
The Help                                                             67   
All the Light We Cannot See                                          67   
A Game of Thrones (A Song of Ice and Fire, #1)                       67   
East of Eden                                                         67   
Unbroken: A World War II Story of Survival, Res...                   67   
The Nightingale                                                      67   

                                                    book_id_consequents  \
consequents                                                               
Harry Potter and the Half-Blood Prince (Harry P...                32805   
Harry Potter and the Deathly Hallows (Harry Pot...                30375   
The Fellowship of the Ring (The Lord of the Rin...                  133   
The Book Thief                                                       94   
The Help                                                             31   
All the Light We Cannot See                                         143   
A Game of Thrones (A Song of Ice and Fire, #1)                       39   
East of Eden                                                        225   
Unbroken: A World War II Story of Survival, Res...                  144   
The Nightingale                                                     267   

                                                    antecedent support  \
consequents                                                              
Harry Potter and the Half-Blood Prince (Harry P...            8.366844   
Harry Potter and the Deathly Hallows (Harry Pot...            8.581521   
The Fellowship of the Ring (The Lord of the Rin...            0.079973   
The Book Thief                                                0.063913   
The Help                                                      0.048682   
All the Light We Cannot See                                   0.048682   
A Game of Thrones (A Song of Ice and Fire, #1)                0.048682   
East of Eden                                                  0.048682   
Unbroken: A World War II Story of Survival, Res...            0.048682   
The Nightingale                                               0.048682   

                                                    consequent support  \
consequents                                                              
Harry Potter and the Half-Blood Prince (Harry P...            9.843475   
Harry Potter and the Deathly Hallows (Harry Pot...            9.737325   
The Fellowship of the Ring (The Lord of the Rin...            0.149575   
The Book Thief                                                0.199626   
The Help                                                      0.051023   
All the Light We Cannot See                                   0.070827   
A Game of Thrones (A Song of Ice and Fire, #1)                0.058296   
East of Eden                                                  0.040004   
Unbroken: A World War II Story of Survival, Res...            0.031795   
The Nightingale                                               0.041553   

                                                     support  \
consequents                                                    
Harry Potter and the Half-Blood Prince (Harry P...  4.402384   
Harry Potter and the Deathly Hallows (Harry Pot...  4.223354   
The Fellowship of the Ring (The Lord of the Rin...  0.037772   
The Book Thief                                      0.010622   
The Help                                            0.004753   
All the Light We Cannot 

**Sonuç**

In [42]:
# Burada bulduğumuz sonuç tek bir kitaba/ayrıntıya bakmaktan ziyade genel resime bakıp bu kişinin ilgi alanına giren
# tüm kitapların belirlenmesidir. Ve bunlar arasında en öne çıkanı tavsiye etmektir. 
# Dilenirse(tavsiyemiz de bu yönde) bunlar arasından matrix factorization yaparak en yüksek puan vereceği ve de 
# o kişi ile en alakalı kitap tavsiye etmek çok mantıklıca olacaktır.

### Yeni Bir Kullanıcı Giriş Yapalım

In [43]:
user1=df_bookish_user[4]
user1

28158

In [44]:
okudugu_kitaplar=df_ratings[df_ratings["user_id"]==user1]["book_id"].values

In [45]:
dfz=dfy[dfy["book_id_antecedents"].isin(okudugu_kitaplar)][~(dfy["book_id_consequents"].isin(okudugu_kitaplar))]
dfz=dfz.groupby("consequents").sum().sort_values("support",ascending=False)
dfz

book_id_antecedents  \
consequents                                                               
Harry Potter and the Chamber of Secrets (Harry ...                27945   
Harry Potter and the Sorcerer's Stone (Harry Po...                27945   
The Book Thief                                                     1686   
The Hero of Ages (Mistborn, #3)                                    2862   
All the Light We Cannot See                                         487   
A Thousand Splendid Suns                                            295   
The Way of Kings (The Stormlight Archive, #1)                       581   
Clockwork Angel (The Infernal Devices, #1)                           70   
The Wise Man's Fear (The Kingkiller Chronicle, #2)                  192   
The Mark of Athena (The Heroes of Olympus, #3)                      294   
Towers of Midnight (Wheel of Time, #13)                             960   
The House of Hades (The Heroes of Olympus, #4)                      294   

                                                    book_id_consequents  \
consequents                                                               
Harry Potter and the Chamber of Secrets (Harry ...                27945   
Harry Potter and the Sorcerer's Stone (Harry Po...                 2430   
The Book Thief                                                      564   
The Hero of Ages (Mistborn, #3)                                    3618   
All the Light We Cannot See                                         572   
A Thousand Splendid Suns                                            201   
The Way of Kings (The Stormlight Archive, #1)                      1124   
Clockwork Angel (The Infernal Devices, #1)                          366   
The Wise Man's Fear (The Kingkiller Chronicle, #2)                  307   
The Mark of Athena (The Heroes of Olympus, #3)                      428   
Towers of Midnight (Wheel of Time, #13)                             988   
The House of Hades (The Heroes of Olympus, #4)                      684   

                                                    antecedent support  \
consequents                                                              
Harry Potter and the Chamber of Secrets (Harry ...            9.383300   
Harry Potter and the Sorcerer's Stone (Harry Po...           10.911422   
The Book Thief                                                0.379339   
The Hero of Ages (Mistborn, #3)                               0.064958   
All the Light We Cannot See                                   0.187131   
A Thousand Splendid Suns                                      0.149323   
The Way of Kings (The Stormlight Archive, #1)                 0.061177   
Clockwork Angel (The Infernal Devices, #1)                    0.109319   
The Wise Man's Fear (The Kingkiller Chronicle, #2)            0.037808   
The Mark of Athena (The Heroes of Olympus, #3)                0.008570   
Towers of Midnight (Wheel of Time, #13)                       0.005977   
The House of Hades (The Heroes of Olympus, #4)                0.008570   

                                                    consequent support  \
consequents                                                              
Harry Potter and the Chamber of Secrets (Harry ...            8.808800   
Harry Potter and the Sorcerer's Stone (Harry Po...            6.065534   
The Book Thief                                                1.197753   
The Hero of Ages (Mistborn, #3)                               0.032119   
All the Light We Cannot See                                   0.283307   
A Thousand Splendid Suns                                      0.146046   
The Way of Kings (The Stormlight Archive, #1)                 0.038888   
Clockwork Angel (The Infernal Devices, #1)                    0.061717   
The Wise Man's Fear (The Kingkiller Chronicle, #2)            0.012315   
The Mark of Athena (The Heroes of Olympus, #3)                0.008354   
Towers of Midnight (Wheel

### Yeni Bir Kullanıcı Giriş Yapalım

In [46]:
user2=df_bookish_user[1]
user2

30944

In [47]:
okudugu_kitaplar=df_ratings[df_ratings["user_id"]==user2]["book_id"].values

In [48]:
dfz=dfy[dfy["book_id_antecedents"].isin(okudugu_kitaplar)][~(dfy["book_id_consequents"].isin(okudugu_kitaplar))]
dfz=dfz.groupby("consequents").sum().sort_values("support",ascending=False)
dfz

book_id_antecedents  \
consequents                                                               
A Dance with Dragons (A Song of Ice and Fire, #5)                  2556   
The Book Thief                                                     1241   
The Titan's Curse (Percy Jackson and the Olympi...                 1002   
The Hero of Ages (Mistborn, #3)                                    1167   
The Well of Ascension (Mistborn, #2)                               1167   
The Help                                                             58   
All the Light We Cannot See                                         231   
The Way of Kings (The Stormlight Archive, #1)                       581   
A Thousand Splendid Suns                                             39   
Clockwork Angel (The Infernal Devices, #1)                           39   
The Hunger Games (The Hunger Games, #1)                              39   

                                                    book_id_consequents  \
consequents                                                               
A Dance with Dragons (A Song of Ice and Fire, #5)                  5076   
The Book Thief                                                      376   
The Titan's Curse (Percy Jackson and the Olympi...                 1284   
The Hero of Ages (Mistborn, #3)                                    1809   
The Well of Ascension (Mistborn, #2)                               1695   
The Help                                                             62   
All the Light We Cannot See                                         286   
The Way of Kings (The Stormlight Archive, #1)                      1124   
A Thousand Splendid Suns                                             67   
Clockwork Angel (The Infernal Devices, #1)                          183   
The Hunger Games (The Hunger Games, #1)                               1   

                                                    antecedent support  \
consequents                                                              
A Dance with Dragons (A Song of Ice and Fire, #5)             0.513827   
The Book Thief                                                0.223282   
The Titan's Curse (Percy Jackson and the Olympi...            0.045585   
The Hero of Ages (Mistborn, #3)                               0.049366   
The Well of Ascension (Mistborn, #2)                          0.049366   
The Help                                                      0.085698   
All the Light We Cannot See                                   0.096104   
The Way of Kings (The Stormlight Archive, #1)                 0.061177   
A Thousand Splendid Suns                                      0.058296   
Clockwork Angel (The Infernal Devices, #1)                    0.058296   
The Hunger Games (The Hunger Games, #1)                       0.058296   

                                                    consequent support  \
consequents                                                              
A Dance with Dragons (A Song of Ice and Fire, #5)             0.484769   
The Book Thief                                                0.798502   
The Titan's Curse (Percy Jackson and the Olympi...            0.038636   
The Hero of Ages (Mistborn, #3)                               0.016671   
The Well of Ascension (Mistborn, #2)                          0.016815   
The Help                                                      0.102045   
All the Light We Cannot See                                   0.141653   
The Way of Kings (The Stormlight Archive, #1)                 0.038888   
A Thousand Splendid Suns                                      0.048682   
Clockwork Angel (The Infernal Devices, #1)                    0.030858   
The Hunger Games (The Hunger Games, #1)                       0.035035   

                                                     support  \
consequents                                                    
A Dance with Dragons (A Song of Ice and Fire, 

### "Kullanıcının ARL mantığına göre şu kitaba  en yakın kitabı bul" isteğini aşağıdaki gibi yapabiliriz

In [49]:
product_id = 19
check_id(df, product_id)

['The Fellowship of the Ring (The Lord of the Rings, #1)']


In [50]:
product_name= "The Fellowship of the Ring (The Lord of the Rings, #1)"

In [51]:
product_name=frozenset({"The Fellowship of the Ring (The Lord of the Rings, #1)"})

In [52]:
recommendation_list=sorted_rules[sorted_rules["antecedents"]==product_name]#.iloc[:3,1]#.values
recommendation_list

antecedents  \
256  (The Fellowship of the Ring (The Lord of the R...   
254  (The Fellowship of the Ring (The Lord of the R...   
683  (The Fellowship of the Ring (The Lord of the R...   
42   (The Fellowship of the Ring (The Lord of the R...   
228  (The Fellowship of the Ring (The Lord of the R...   
253  (The Fellowship of the Ring (The Lord of the R...   
250  (The Fellowship of the Ring (The Lord of the R...   

                                           consequents  \
256       (The Two Towers (The Lord of the Rings, #2))   
254  (The Return of the King (The Lord of the Rings...   
683  (The Return of the King (The Lord of the Rings...   
42    (A Game of Thrones (A Song of Ice and Fire, #1))   
228                                   (The Book Thief)   
253  (The Lord of the Rings (The Lord of the Rings,...   
250                                         (The Help)   

     antecedent support  consequent support  \
256            0.027402            0.011594   
254            0.027402            0.011882   
683            0.027402            0.008894   
42             0.027402            0.058296   
228            0.027402            0.099813   
253            0.027402            0.015231   
250            0.027402            0.051023   

      support  confidence       lift  leverage  \
256  0.006373    0.232589  20.060414  0.006056   
254  0.006193    0.226018  19.021160  0.005868   
683  0.005581    0.203679  22.901148  0.005337   
42   0.003889    0.141919   2.434442  0.002291   
228  0.003493    0.127464   1.277030  0.000758   
253  0.002881    0.105125   6.901955  0.002463   
250  0.002521    0.091984   1.802813  0.001122   

     conviction  
256    1.287974  
254    1.276668  
683    1.244607  
42     1.097453  
228    1.031690  
253    1.100454  
250    1.045111